fdbserver/ClusterController.actor.cpp

```diff
@@ -2274,7 +2275,9 @@ ACTOR Future<Void> startEncryptKeyProxy(ClusterControllerData* self) {
 			                                                                       self->db.config,
 			                                                                       id_used);
 
+			state EncryptionAtRestMode encryptMode = wait(self->encryptionAtRestMode.getFuture());
 			InitializeEncryptKeyProxyRequest req(deterministicRandom()->randomUniqueID());
+			req.encryptMode = encryptMode;
```

Comment:
why is this needed? Am thinking if change would cause any deadlocks during startup? As proposal EKP would wait for DB config to be available?

Reply:
Alternatively, EKP process can be started regardless, however, the encryption endpoints needs to check if `encryptionAtRestMode != disabled`

Reply:
It's not strictly needed. I added some code probes and asserts in the EKP that use the encrypt mode to make sure that we aren't accessing encryption endpoints if encryption is disabled but the knob is enabled. But we call this in `startEncryptKeyProxy`  which itself is done after reading the value of the encryptMode from the DB config so we don't need to worry about deadlock cases

Reply:
> Alternatively, EKP process can be started regardless, however, the encryption endpoints needs to check if encryptionAtRestMode != disabled

In order for the EKP to check the encrypt mode we need to pass it here

Reply:
We can optimize the recruitment of this by not waiting for the db config if the knob is enabled but again im not sure how much overhead that would reduce and it would then be harder to see fi we are accessing encryption endpoints when the encryption is off because the db config would not yet be ready

Reply:
My recommendation is, if it is not strictly needed, better not to add any dependency. Things might work today, but then any future changes might add deadlocks, hence, code becomes brittle. Especially if the usage is only to support `asserts`, it might be better to find alternatives to achieve the same.

For instance: EKP today takes `db` object which is itself not used. 
`encryptKeyProxyServer(EncryptKeyProxyInterface ekpInterface, Reference<AsyncVar<ServerDBInfo>> db)`, it can be used to interact with CC interface to fetch `encryptMode` in simulation and support necessary asserts.

Reply:
TLog uses this method to get encryptionAtRestMode already

```
ACTOR Future<EncryptionAtRestMode> getEncryptionAtRestMode(TLogData* self) {
	loop {
		state GetEncryptionAtRestModeRequest req(self->dbgid);
		try {
			choose {
				when(wait(self->dbInfo->onChange())) {}
				when(GetEncryptionAtRestModeResponse resp = wait(brokenPromiseToNever(
				         self->dbInfo->get().clusterInterface.getEncryptionAtRestMode.getReply(req)))) {
					TraceEvent("GetEncryptionAtRestMode", self->dbgid).detail("Mode", resp.mode);
					return (EncryptionAtRestMode::Mode)resp.mode;
				}
			}
		} catch (Error& e) {
			TraceEvent("GetEncryptionAtRestError", self->dbgid).error(e);
			throw;
		}
	}
}
```
Again, it can be avoided for production code and kept strictly for simulation runs

Reply:
The TLOG example above ends up waiting on the same call we will (i.e for the db config to be ready). The `db` object that reaches out to the CC also uses the exact same method that we use here. The way i did it above is how we do it for other components like the CP, SS, Resolver. If you look at `InitializeCommitProxyRequest` etc.. this is the standard way which we discussed with Evan. I would rather keep it this way than try to do something unique for each component, it makes it harder to understand the code

Reply:
This is the earliest point in which we have the DBConfig available so not really sure if there is any more optimizations we can do

Reply:
My suggestion was bit different (above code was an example how to do it)
1. Let EKP start without adding any additional depedency
2. Define `Optional<EncryptionAtRest>` config in the EKP state
3. In Simulation, on call to fetch encryption keys, check if `optional` is set, if not, use above method to get the encryptionMode. 
4. Again, once the mode is fetched, all asserts should work fine.